### PR TITLE
Third-order Runge-Kutta timestepper

### LINKED ISSUE #506: Forward Euler + 3rd order Runge-Kutta time-stepping (or integration with DifferentialEquations.jl)
We need to implement a forward Euler time-stepping scheme, which is useful for testing if for nothing else.

We need also better and more stable time-steppers. One option is to implement by hand a 3rd-order low storage Runge Kutta scheme.

Another interesting avenue is to figure out how to integrate with `DifferentialEquations.jl`, which would give us access to a large number of new time-steppers (issue #391). However, we are unsure how this would couple with the our incoming GPU tridiagonal solvers for implicit time-stepping.

### PR DESCRIPTION
**Pull request by glwagner**
This PR implements a third-order Runge-Kutta time stepping method.

The implementation is based off [Le and Moin (1991)](https://www.sciencedirect.com/science/article/pii/0021999191902157). It should be noted, however, that we do not use an implicit method for diffusion, and that the pressure correction is calculated every substep. It could be possible in the future to implement the optimization described in Le and Moin (1991) that reduces the number of times the pressure Poisson equation needs to be solved each time-step from 3 to 1, though this would reduce the accuracy of the scheme from third to second-order.

So far, this PR extends the dynamics tests and incompressibility test to `RungeKutta3TimeStepper`. It also extends the time stepper convergence test to the `RungeKutta3TimeStepper`.

I am open to changing the name of the time stepper. It may also be a good idea to add basic documentation. The time stepper is used by setting the keyword `timestepper=:RungeKutta3` in the constructor for `IncompressibleModel`.

~~I am not sure if checkpointing will work with this timestepper.~~

Checkpointing will not work with this timestepper. A more generic checkpointer awaits a future PR.

Resolves #506

### CODE REVIEW

**Comment by glwagner** [comment]
This plot could be spruced up a bit, but its showing the right thing:

image

**Comment by glwagner** [comment]
@ali-ramadhan any more tests you can think of?

**Review by ali-ramadhan — APPROVED** [review]
Finally! :tada:

Looks great. Just wondering whether RK3 should become the default time-stepper in this PR or in a later PR?

Also, I think you split the tests on Travis only so GitLab CI is missing Linux + GPU tests for `TEST_GROUP={model,simulation}`.

PS: This PR resolves #506

File: src/TimeSteppers/rk3.jl [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    precomputations!(diffusivities, pressures, velocities, tracers, model)
+
+    calculate_tendencies!(Gⁿ, velocities, tracers, pressures, diffusivities, model)
+
+    rk3_substep!(velocities, tracers, arch, grid, Δt, γ¹, nothing, Gⁿ, nothing)
+
+    calculate_pressure_correction!(pressures.pNHS, first_substep_Δt, velocities, model)
+    pressure_correct_velocities!(velocities, arch, grid, first_substep_Δt, pressures.pNHS)
+
+    tick!(model.clock, first_substep_Δt)
```
[the remark this answers is not in the document]
Comment (glwagner) [reply src/TimeSteppers/rk3.jl]: That's a good point. `substep` would be needed if implementing a stochastic forcing with a random component, which should only be updated every timestep and left constant during substeps.

[the remark this answers is not in the document]
Comment (glwagner) [reply src/TimeSteppers/rk3.jl]: I think the common word is `stage`, rather than `substep`

[the remark this answers is not in the document]
Comment (glwagner) [reply src/TimeSteppers/rk3.jl]: I've added a field to `clock` called `clock.stage`, and a keyword argument to `tick!`; `stage=true` means "advance the clock stage`, whereas `stage=false` means "advance the clock iteration and reset the stage".

**Comment by glwagner** [comment]
> Just wondering whether RK3 should become the default time-stepper in this PR or in a later PR?

I think once we have convergence tests that use `RungeKutta3` we are ready to make it the default? We also need to add new regression tests.

**Pull request merged**